### PR TITLE
dialect: (riscv) initial custom syntax

### PR DESCRIPTION
**Pull request by kingiler**
This is an initial PR to custom RISCV syntax.

The major change is dropping bracket around operands

### CODE REVIEW

File: xdsl/dialects/riscv.py [hunk 1]
```diff
@@ -346,13 +351,55 @@ def parse(cls, parser: Parser) -> Self:
             regions=regions,
         )
 
+    @classmethod
+    def parse_args(cls, parser: Parser) -> Sequence[UnresolvedOperand]:
+        if arg := parser.parse_optional_unresolved_operand():
```
Comment (superlopuh) [inline xdsl/dialects/riscv.py:356]: I think @math-fehr added an optional list without delimiter helper recently, could you please take a look?

File: xdsl/dialects/riscv.py [hunk 2]
```diff
@@ -346,13 +351,55 @@ def parse(cls, parser: Parser) -> Self:
...
+        return []
+
+    @classmethod
+    def parse_attributes(cls, parser: Parser) -> Mapping[str, Attribute]:
+        return parser.parse_optional_attr_dict()
+
     def print(self, printer: Printer) -> None:
-        printer.print_operands(self.operands)
+        first = True
+        first &= not self.print_operands(printer, first)
```
Comment (superlopuh) [inline xdsl/dialects/riscv.py:371]: what does first mean? it feels like this logic could be shifted out of the helper methods, and be here for clarity, even if that means checking for an empty sequence twice

File: xdsl/dialects/riscv.py [hunk 3]
```diff
@@ -346,13 +351,55 @@ def parse(cls, parser: Parser) -> Self:
...
+            if not first:
+                printer.print(",")
+            printer.print(" ")
+            printer.print_list(self.operands, printer.print_operand)
+        return self.operands
+
+    def print_attributes(self, printer: Printer, first: bool) -> Set[str]:
+        """
+        Print list of attributes,
+        return the list of attributes being processed.
```
Comment (superlopuh) [inline xdsl/dialects/riscv.py:398]: what does this mean? can you please elaborate in the comment? it feels like this is another case of adding helper methods before it is clear that they will be used later on. Adding a `self.print_attributes` method that just forwards to priner.print_op_attributes makes the code less clear in the short term. Am I missing something? Is it necessary?

**Review by superlopuh — CHANGES_REQUESTED** [review]

File: xdsl/dialects/riscv.py (hunk 3 above)
Comment (tobiasgrosser) [reply xdsl/dialects/riscv.py:398]: I would suggest to first put up a pr that just drops the `()` around all operands. This should be a one-line change and does not yet need any of the helper functions. Let's avoid introducing all the helper functions, as these are controversial and I would like to discuss  them with Anton in person.

Comment (superlopuh) [reply xdsl/dialects/riscv.py:398]: Ah no I see it is being overridden below, then please expand on the comment on how the logic works. To me, it seems like the list of attributes returned is not "processed" but rather "printed", maybe "printed in the custom format". I'd recommend changing the name of the method to something like "print_known_attributes" or, "print_attributes_custom_format"

File: xdsl/dialects/riscv.py (hunk 1 above)
Comment (kingiler) [reply xdsl/dialects/riscv.py:356]: It seems there are some issues with `parser.parse_optional_undelimited_comma_separated_list`.

Comment (math-fehr) [reply xdsl/dialects/riscv.py:356]: This function should be put directly in the `parser` I would argue.
Also, what issue do you have with this function? `parser.parse_optional_undelimited_comma_separated_list(parser.parse_optional_unresolved_operand, parser.parse_unresolved_operand)` should work without problems.

File: xdsl/dialects/riscv.py [hunk 4]
```diff
@@ -346,13 +351,55 @@ def parse(cls, parser: Parser) -> Self:
...
+            args = [arg]
+            while parser.parse_optional_punctuation(",") and (
+                arg := parser.parse_optional_unresolved_operand()
+            ):
+                args.append(arg)
+            return args
+        return []
+
+    @classmethod
+    def parse_attributes(cls, parser: Parser) -> Mapping[str, Attribute]:
```
Comment (math-fehr) [inline xdsl/dialects/riscv.py:366]: Why do we redefine these things here? Can we just call `parser.parse_optional_attr_dict` directly?

File: xdsl/dialects/riscv.py [hunk 5]
```diff
@@ -346,13 +351,55 @@ def parse(cls, parser: Parser) -> Self:
...
+            for name, attr in self.attributes.items()
+            if name not in processed_attrs
+        }
+        printer.print_op_attributes(remaining_attrs)
         printer.print_regions(self.regions)
-        printer.print_op_attributes(self.attributes)
         printer.print(" : ")
         printer.print_operation_type(self)
 
+    def print_operands(self, printer: Printer, first: bool) -> Sequence[SSAValue]:
```
Comment (math-fehr) [inline xdsl/dialects/riscv.py:383]: Can we move that in the printer as well?

File: xdsl/dialects/riscv.py (hunk 1 above)
Comment (kingiler) [reply xdsl/dialects/riscv.py:356]: In the new custom syntax, both `args` and attributes are comma separated.

`riscv.op %0, %1, 123, "comment" (regions) ...`

File: xdsl/dialects/riscv.py (hunk 4 above)
Comment (kingiler) [reply xdsl/dialects/riscv.py:366]: `parse_attributes` will be override by classes which implement custom attribute syntax. By default it uses generic attribute dict. See [here](https://github.com/xdslproject/xdsl/pull/1318/files#diff-c2e583c286e17ace72eadfcf055304a684874cd8bfbb4c2b51ff7d9d21ff7be4R2141) for `riscv.DirectiveOp` example.

**Comment by kingiler** [comment]
Now the PR is minimized to only drop brackets around operands.

**Review by tobiasgrosser — APPROVED** [review]
I guess this looks good to me so far.

**Comment by tobiasgrosser** [comment]
@superlopuh, could you have a last look?

File: xdsl/dialects/riscv.py [hunk 6]
```diff
@@ -333,7 +333,13 @@ def assembly_line(self) -> str | None:
...
     @classmethod
     def parse(cls, parser: Parser) -> Self:
-        args = parser.parse_op_args_list()
+        args = (
+            parser.parse_optional_undelimited_comma_separated_list(
+                parser.parse_optional_unresolved_operand,
+                parser.parse_unresolved_operand,
+            )
+            or ()
+        )
```
Comment (superlopuh) [inline xdsl/dialects/riscv.py:342]:
```suggestion
        args = parser.parse_optional_undelimited_comma_separated_list(
            parser.parse_optional_unresolved_operand,
            parser.parse_unresolved_operand,
        )
        if args is None:
            args = ()
```

We try not to exploit the falseyness of `None` due to the ambiguity of double falseyness. Sometimes the result of some function is not `None` but still falsey, which can lead to hard-to-debug behaviour.

**Review by superlopuh — CHANGES_REQUESTED** [review]
LGTM modulo small change

File: xdsl/dialects/riscv.py (hunk 6 above)
Comment (kingiler) [reply xdsl/dialects/riscv.py:342]: `parser.parse_optional_undelimited_comma_separated_list` returns either list or `None`. The return value is falsy only if it is empty list or `None` and either case returning an empty tuple is okay.

**Pull request merged**